File: time.rb
class TimeCode
  def initialize(count = 4, note = 4)
    @at = {}
    @clock = 0

    @beat_len = 96 / note
    @bar_len  = @beat_len * count
  end

  def pulse
    unless @at[@clock].nil? 
      @at[@clock].call
    end
    @clock += 1
  end

#  def reset 
#    @clock = 0
#  end

#  alias_method :stop,  :reset
#  alias_method :start, :reset

  def to_s
    pos  = @clock % @bar_len + 1
    bar  = @clock / @bar_len + 1
    beat = @clock % @bar_len / @beat_len + 1
    "#{bar} : #{beat} : #{pos}"
  end

  def at(bar: 1, beat: 1, &block)
    p = ((bar - 1) * @bar_len) + ((beat - 1) * @beat_len)
    @at[p] = block
  end
end

File: rack/time.js
class Time {
    constructor(count = 4, note =4) {
        this.events  = {};
        this.run     = false;
        this.steps   = 0;
        this.beatLen = 96 / note;
        this.barLen  = this.beatLen * count;
    }

    step() {
        if(this.run) {
            Object.keys(this.events).map((k) => {
                if(parseInt(k) === this.steps) {
                    this.events[k]();
                }
            });
            this.steps += 1;
        }
    }

    log(message = "") {
        let pos  = this.steps % this.barLen + 1;
        let bar  = Math.floor(this.steps / this.barLen) + 1;
        let beat = Math.floor(this.steps % this.barLen / this.beatLen) + 1
        console.log(`${bar} : ${beat} : ${pos} ${message}`) 
    }

    start() {
        this.run = true;
    }

    stop() {
        this.run   = false;
        this.steps = 0;
    }

    at({bar = 1, beat = 1}, func) {
        let t = ((bar - 1) * this.barLen) + ((beat - 1) * this.beatLen)
        this.events[t] = func;
    }
}

module.exports = Time;
